src/rust/engine/src/interning.rs

```diff
@@ -43,14 +43,32 @@ impl Interns {
     Interns::default()
   }
 
+  pub fn insert_product(&mut self, v: Value) -> Key {
```

Comment:
Does this need to be a separate method, or could just using `insert` work? AFAIK, the `type(some_type)` should just be `type`, so calling `insert` should work.

Reply:
This was added in a hurry when I didn't know why `identify()` was causing errors, this edit sounds right.

Reply:
Well, we do want to make sure the type is stored in `self._types` in `native.py`, right? Are you thinking of using `insert` without removing `extern_product_type()` somehow?

Reply:
Hm. Should the generator code (in https://github.com/pantsbuild/pants/blob/3a81e86eca33bcd16ca68838052ffeb25fdcc803/src/python/pants/engine/native.py#L417-450) memoize these in `self._types` before returning them?

Reply:
I was also thinking that we could add another argument to `extern_identify()` which says to consider the handle as a `type` and intern that instead of taking the `type()` of the handle.

But I think the solution you're suggesting here with memoizing in `extern_generator_send()` sounds like we would end up being able to remove all of the ffi calls in `interning.rs`? Because that sounds pretty neat.

Reply:
I've pulled this thread in another PR: should have it out in a few minutes.

Reply:
#7318 

Regarding completely eliminating the need to call back to python in `interning.rs`: I think that you could almost do it, but we'd probably need to make the `PyGeneratorResponse` type more complicated, because we only hash the keys in `Gets`, and not in `Breaks`.